go/private/extensions.bzl

```diff
@@ -101,6 +112,42 @@ def _go_sdk_impl(ctx):
                 sdk_version = download_tag.version,
             ))
 
+            # Additionally register SDKs for all common execution platforms, but only if the user
+            # specified a version to prevent eager fetches.
+            if download_tag.version and not download_tag.goos and not download_tag.goarch:
+                for goos, goarch in _COMMON_EXEC_PLATFORMS:
+                    if goos == host_detected_goos and goarch == host_detected_goarch:
+                        # We already added the host-compatible toolchain above.
+                        continue
+
+                    if download_tag.sdks and not "{}_{}".format(goos, goarch) in download_tag.sdks:
+                        # The user supplied custom download links, but not for this tuple.
+                        continue
+
+                    default_name = _default_go_sdk_name(
+                        module = module,
+                        multi_version = multi_version_module[module.name],
+                        tag_type = "download",
+                        index = index,
+                        suffix = "_{}_{}".format(goos, goarch),
+                    )
+                    go_download_sdk_rule(
+                        name = default_name,
+                        goos = download_tag.goos,
+                        goarch = download_tag.goarch,
+                        sdks = download_tag.sdks,
+                        urls = download_tag.urls,
+                        version = download_tag.version,
+                    )
+
+                    toolchains.append(struct(
+                        goos = goos,
+                        goarch = goarch,
+                        sdk_repo = default_name,
+                        sdk_type = "remote",
+                        sdk_version = download_tag.version,
+                    ))
+
```

Comment:
would this still work if you move the `go_download_sdk_rule` above this `if` clause into an `else` clause?

i.e.
```
            if download_tag.version and not download_tag.goos and not download_tag.goarch:
                for goos, goarch in _COMMON_EXEC_PLATFORMS:
                    ...
                    go_download_sdk_rule(...)
                    ...
            else:
                ...
                go_download_sdk_rule(...)
                ...
```

Reply:
It makes a difference if the host SDK is not contained in the list of common exec platforms. We always want to register an SDK for the host, even if it is not in the list.

Reply:
So with the current approach, there's potentially be an SDK registered twice (with different names).

I don't think it will be a big deal as the final Merkle tree hash would still be the same regardless. But it could be a confusing UX for the end users if both ended up being used somehow. 🤔

Reply:
Toolchain registration order definitely matters. I would think that "all 1.19 SDKs with varying exec platforms" coming before "all 1.20 SDKs with varying exec platforms" would be the most reasonable behavior, but of course this PR could end up changing which SDK is selected for existing declarations of the form:

```starlark
go_sdk.download(version = "1.19")
go_sdk.download(version = "1.20", goos = "linux", goarch = "arm64")
```

On a Linux amd64 host with Linux arm64 executor, with this PR 1.19 would be selected instead of the previous selection of 1.20. I don't know how common this setup is in practice, it does look error-prone as is.

Re duplicate SDKs: Unfortunately the Merkle tree hash includes the repo name, meaning that currently cache hits can fail for a variety of reasons, including just bumping the rules_go version (which is part of the repo name) or registering an additional toolchain earlier in the module even if it doesn't end up being selected. The ongoing path mapping work may help with that some day, but it's a difficult problem to solve in general.

Reply:
I would like to revive this PR as I think it's still very relevant.

> I think as long as we provide a knob for people to opt-out of the default automatic toolchain registering to use their own, then they could do w/e they want. So what do you think we just add a flag (rules / extension attribute) in to disable the auto toolchain registration for known platforms?

I would like to do away with the "registering toolchains is costly" point of view that IMO has historically plagued rulesets. As part of this, I would prefer not viewing registering additional toolchains as a breaking change. How about we assume that the current mechanisms for controlling resolution (e.g. `go_cross_binary`) are sufficient for now and if really turns out that they aren't, we improve user control over resolution instead of making registration more complicated? I can see that being more maintainable down the road and also more useful for users.